### DISCUSSION TITLE
Save input values

### DISCUSSION THREAD

**Discussion by Alperenn53** [bitcraze/discussions#501] [Q&A]
How can I save push, M1, M2, M3 and M4 input values second by second from crazyflie 2.1 interface?

**Comment by hmllr** [comment]
There is a logging/parameter framework, you can add the variables you want to log to it following the examples in [https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/userguides/logparam/](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/userguides/logparam/) and then log them over the cfclient or even on the SDcard.

Hope this helps,
Hanna

**Reply by Alperenn53** [reply]
Thank you. So what is the data unit I get when I record the m1, m2, m3, m4 and impulse graphs second by second? (Voltage, RPM etc.) What kind of data is this? Do you know this?

**Reply by hmllr** [reply]
this is the PWM ratio, the maximum is an uint16 (65535). So the M1, M2, M3, M4 values are relative to the voltage applied to the motors.
Does this help?

**Reply by Alperenn53** [reply]
Thank you so much.

**Reply by Alperenn53** [reply]
![Ekran Görüntüsü (134)](https://user-images.githubusercontent.com/111359552/211781102-9eb661bf-4dec-4e9d-9e0e-f356352757ba.png)

So the values I draw from the motor are PWM signals? If yes, where can I find out how much voltage I can reach these values? Is there an equation link between PWM signals and voltages?

**Reply by hmllr** [reply]
where are you logging those variables? They seem to be much higher than uint16. Actually, you could use the ones already logged in the stabilizer, they are uncapped as well but on the right scale. https://github.com/bitcraze/crazyflie-firmware/blob/master/src/modules/src/stabilizer.c#L818
The voltage then has a linear relation - so just divide by 65535 and multiply by your battery voltage.

**Comment by Alperenn53** [comment]
I was surprised that they were higher. I save these values directly from the plotter of the Crazyflie software interface. I see the values in Visual Studio Code, but the values are too large.

**Reply by Alperenn53** [reply]
Can you explain step by step to solve this problem and save motor inputs? Later, I will find out what types of these inputs are and integrate them into my model.

**Reply by hmllr** [reply]
if you log them directly from the plotter it should be fine actually - I think the issue is that you are not actually flying, right? Then the motors might saturate, and as this logging variable is before the capping it is normal to get values that seem impossible. Try it during a normal flight and see if it looks more reasonable.

**Reply by Alperenn53** [reply]
I am getting these values during 'Hovering'. Am I pulling the wrong values?

**Reply by hmllr** [reply]
which values are you printing exactly? From your screenshot it seems that they are named motor.m4reg_x and so on, but the values in the stabilizer module should be called motor.m4req and so on? they don't have an x and y component.

**Reply by Alperenn53** [reply]
I want to get the inputs of the motor in full. These are Thrust, m1, m2, m3, and m4. I need to learn how these values are formed. You mentioned a formula in the article you wrote above. Is there a clearer version of this formula? If so, how do I get to the battery voltage there, second by second? In short, I will model the crazyflie input values in Matlab/Simulink. Should I supply voltage to the inputs or get a PWM signal? If I'm going to receive a PWM signal, what does that PWM signal depend on? How can I pull the parameters that the PMW signal is connected to from the interface?

**Reply by hmllr** [reply]
the inputs for the motors are only m1, m2, m3 and m4. Those are computed from the pitch/yaw/roll/thrust outputs of the PID controller (https://github.com/bitcraze/crazyflie-firmware/blob/master/src/modules/src/power_distribution_quadrotor.c#L79), with applying a battery compensation and capping them to uint16 (https://github.com/bitcraze/crazyflie-firmware/blob/master/src/modules/src/stabilizer.c#L309). There are plenty of tutorials in the web about how to drive brushed DC motors with PWM signals, as for example: https://learn.adafruit.com/improve-brushed-dc-motor-performance/pwm-and-brushed-dc-motors 
The PWM signal is directly proportional to how strong you want to turn on a specific motor and it should be possible to log them with the motor.mXreg variables, where X is 1-4. Those values are uncapped, but you can easily cap them to uint16 after logging as well.
Does this answer your question?

**Reply by Alperenn53** [reply]
Thank you so much. I got the answer to my question :)
